allow HTTP headers to be supplied for outcome request

hmoffatt opened the pull request:
I need to be able to override the User-Agent in the POST request because an outcome service I'm reporting to is rejecting the default python-requests user agent.

hmoffatt committed:
Allow HTTP headers to be supplied for outcome request

ryanhiebert reviewed (changes requested):
This looks like a generally good change, thank you for bringing it up! There are a couple things that need to be addressed before we can merge this.

First, using `headers={}` as defaults in function signatures is a bad idea, because that `dict` instance is preserved and mutable. See https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments for more on this.

Second, it needs some tests. I don't use this feature in my own work right now, so its that much more important that there are good automated tests. This seems like it may be a less-tested part of the code, but I'll be satisfied when the lines added here are covered.

Finally, should the headers be passed in for each individual method, or would it be best to have them in the `__init__` and `from_post_request` constructors? Since your use-case is about overriding the user agent in particular, it seems like the class-level might be the better place for it.

hmoffatt commented:
Thanks for the feedback! I have addressed all three issues in my latest changes.

hmoffatt commented:
Did you have a chance to review this @ryanhiebert ?

ryanhiebert commented on src/lti/outcome_request.py:
```diff
@@ -51,15 +52,19 @@ def __init__(self, opts=defaultdict(lambda: None)):
                     "Invalid outcome request option: {}".format(key)
                 )
 
+        self.headers = CaseInsensitiveDict(headers)
```
Since `headers` may be `None`, will this actually work? It's not typical for that to work for `dict`. I'd probably prefer this to be written:

```suggestion
        self.headers = CaseInsensitiveDict(headers or {})
```

hmoffatt replied:
requests' CaseInsensitiveDict handles that case, but I've added the code you suggested for clarity.

hmoffatt replied:
I've extended the test a bit further so there's a net improvement in the test coverage.

ryanhiebert reviewed (commented):
Thanks for adding some tests there. I'm not sure what's going on with the tests, but I'm going to need to get that figured out before I'm ready to merge.

hmoffatt committed:
Add support for Canvas ltiLaunchUrl in outcome request

ryanhiebert commented:
Hmm. It seems like CI isn't doing anything. I'll need to see what I can do about that.

hmoffatt commented:
Sorry, accidentally pushed an unrelated change. I have removed it and also squashed my previous commits down to one.

hmoffatt committed:
Merge branch 'ltiLaunchUrl' into production

hmoffatt committed:
parse OutcomeRequest result data

hmoffatt committed:
add support for Canvas needsAdditionalReview flag
https://github.com/instructure/canvas-lms/blob/master/doc/api/assignment_tools.md#submission-needs-additional-review

hmoffatt committed:
allow a requests session to be passed in to outcome reporter
